scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+        exceeded, the algorithm will stop just after the local search is
+        done.
+    args: Optional arguments to be passed to the objective function call.
+
+
+    Returns
+    -------
+    res : OptimizeResult
+        The optimization result represented as a ``OptimizeResult`` object.
+        Important attributes are: ``x`` the solution array, ``fun`` the value
+        of the function at the solution, and ``message`` which describes the
+        cause of the termination.
+        See `OptimizeResult` for a description of other attributes.
+
+    Notes
+    -----
+    GenSA is an implementation of the General Simulated Annealing algorithm
+    (GSA [2]_). This stochastic approach generalizes CSA (Classical Simulated
+    Annealing) and FSA (Fast Simulated Annealing) to find the neighborhood of
+    minima, then calls a local method (lbfgs) to find their exact value.
+    The algorithm was originally implemented in C++/R and is explained in
+    more detail by Yang Xiang and al [5]_.
+    GenSA can process complicated and high dimension non-linear objective
+    functions with a large number of local minima as described by Muller
+    paper [6]_.
+    .. versionadded:: 0.18.0
+
+    References
+    ----------
+    .. [1] Tsallis C (1988). "Possible generalization of Boltzmann-Gibbs
+        statistics." Journal of Statistical Physics, 52, 479-487.
+    .. [2] Tsallis C, Stariolo DA (1996). "Generalized Simulated Annealing."
+        Physica A, 233, 395-406.
+    .. [3] Xiang Y, Sun DY, Fan W, Gong XG (1997). "Generalized Simulated
+        Annealing Algorithm and Its Application to the Thomson Model."
+        Physics Letters A, 233, 216-220.
+    .. [4] Xiang Y, Gong XG (2000a). "Efficiency of Generalized Simulated
+        Annealing." PHYSICAL REVIEW E, 62, 4473.
+    .. [5] Xiang Y, Gubian S, Suomela B, Hoeng (2013). "Generalized Simulated
+        Annealing for Efficient Global Optimization: the GenSA Package for
+        R". The R Journal, Volume 5/1, June 2013.
+        http://journal.r-project.org/.
+    .. [6] Mullen, K. (2014). Continuous Global Optimization in R. Journal of
+        Statistical Software, 60(6), 1 - 45.
+        http://dx.doi.org/10.18637/jss.v060.i06
+
+    Examples
+    --------
+    The following example is a 10-dimensional problem, with many local minima.
+    The function involved is called Rastrigin
+    (https://en.wikipedia.org/wiki/Rastrigin_function)
+
+    >>> from scipy.optimize import gensa
+    >>> func = lambda x: np.sum(x * x - 10 * np.cos(2 * np.pi * x))\
+            + 10 * np.size(x)
+    >>> lw = [-5.12] * 10
+    >>> up = [5.12] * 10
+    >>> ret = gensa(func, None, bounds=(zip(lw, up)))
+    >>> print("global minimum: xmin = {0}, f(xmin) = {1}".format(
+    ...    ret.x, ret.fun))
+    """
+    lower, upper = zip(*bounds)
+    gr = GenSARunner(func, x0, np.array(lower), np.array(upper), fargs=args)
+    gr._tempsta = T
```

Comment:
Don't set private variables like this. I would add all these variables in L577-> L583 as keywords to `GenSARunner.__init__`. Can you make `GenSARunner` a class based solver, with `gensa` being a wrapper function for calling `GenSARunner`?